ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveRowIsDeletedPropagator.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.optimizer.calcite.rules.views;
+
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexShuttle;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.hadoop.hive.ql.metadata.VirtualColumn;
+import org.apache.hadoop.hive.ql.optimizer.calcite.HiveRelShuttle;
+import org.apache.hadoop.hive.ql.optimizer.calcite.HiveRelShuttleImpl;
+import org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveJoin;
+import org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveProject;
+import org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveTableScan;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * {@link HiveRelShuttle} to propagate rowIsDeleted column to all HiveRelNodes' rowType in the plan.
+ * General rule: we expect that the rowIsDeleted column is the last column in the input rowType of the current
+ * {@link org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveRelNode}.
+ */
+public class HiveRowIsDeletedPropagator extends HiveRelShuttleImpl {
```

Comment:
Isn't the propagator missing the Filter operator? Could we add an incremental maintenance test for a MV that has a query definition with a Filter on top of the TS with the deleted row?

Reply:
TS operators always has `ROW__IS__DELETED` in their RowType because it is added along with other virtual columns when creating CBO plan from `QB` objects then filters operators derive their RowType from their input.

I tried rebuilding an MV like:
```
CREATE MATERIALIZED VIEW cmv_mat_view_n6 TBLPROPERTIES ('transactional'='true') AS
  SELECT a, c
  FROM cmv_basetable_2_n3
  WHERE NOT ROW__IS__DELETED;
```
and got a correct plan:
```
CBO PLAN:
HiveProject(a=[$1], c=[$2])
  HiveFilter(condition=[OR($3, NOT($3))])
    HiveJoin(condition=[IS NOT DISTINCT FROM($0, $1)], joinType=[right], algorithm=[none], cost=[not available])
      HiveProject(a=[$0])
        HiveTableScan(table=[[default, cmv_mat_view_n6]], table:alias=[default.cmv_mat_view_n6])
      HiveProject(a=[$0], c=[$2], ROW__IS__DELETED=[$7])
        HiveFilter(condition=[AND(<(1, $6.writeid), NOT($7))])
          HiveTableScan(table=[[default, cmv_basetable_2_n3]], table:alias=[cmv_basetable_2_n3])
```

Reply:
I see. What happens if you have an MV with two filters? Would it still work?

```
  SELECT a, c FROM (
    SELECT a, c
    FROM cmv_basetable_2_n3
    WHERE NOT c < 15) subq
  WHERE NOT c < 10;
```

Reply:
This case filters are merged. The Propagator is called after `applyPreJoinOrderingTransforms`.